server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -201,4 +204,82 @@ public boolean equals(Object o) {
     public int hashCode() {
         return Objects.hash(name, timeStampField, indices, generation);
     }
+
+    public static final class TimestampField implements Writeable, ToXContentObject {
+
+        static ParseField FIELD_NAME_FIELD = new ParseField("field_name");
+        static ParseField FIELD_MAPPING_FIELD = new ParseField("field_mapping");
+
+        @SuppressWarnings("unchecked")
+        private static final ConstructingObjectParser<TimestampField, Void> PARSER = new ConstructingObjectParser<>(
+            "timestamp_field",
+            args -> new TimestampField((String) args[0], (Map<String, Object>) args[1])
+        );
+
+        static {
+            PARSER.declareString(ConstructingObjectParser.constructorArg(), FIELD_NAME_FIELD);
+            PARSER.declareObject(ConstructingObjectParser.optionalConstructorArg(), (p, c) -> p.mapOrdered(), FIELD_MAPPING_FIELD);
+        }
+
+        private final String fieldName;
+        private final Map<String, Object> fieldMapping;
+
+        public TimestampField(String fieldName, Map<String, Object> fieldMapping) {
+            this.fieldName = fieldName;
+            this.fieldMapping = fieldMapping;
```

Comment:
Should we assert that the fieldMapping contains ("type")?

Reply:
I added that assert here: https://github.com/elastic/elasticsearch/pull/58096/files#diff-82f348b800c60fabd3a34ff574e248a5R159

The reason is that in production, this is where new data stream instances are initiated.
Ideally this assertion would be added here, but then all tests need to be changes to supply: `Map.of("type, "date")` whereas currently tests supply: `Map.of()`. Maybe it makes sense to add an new constructor that just sets `fieldMapping` to `Map.of("type, "date")` and then delegates to this constructor where we can assert whether `type` key contains a valid value?

Reply:
Could we not just supply `DataStreamTestHelper.defaultMapping()` (new method) instead? A few more bytes of source code, but I would prefer asserting this here to avoid something creating this in a wrong way somewhere or under some condition.